flink-tests/src/test/java/org/apache/flink/test/checkpointing/ImmediateCheckpointingAfterAllTasksFinishedITCase.java

```diff
@@ -0,0 +1,202 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.checkpointing;
+
+import org.apache.flink.api.common.restartstrategy.RestartStrategies;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.api.connector.sink2.Sink;
+import org.apache.flink.api.connector.sink2.SinkWriter;
+import org.apache.flink.core.execution.SavepointFormatType;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.jobgraph.SavepointRestoreSettings;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.runtime.minicluster.MiniClusterConfiguration;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.runtime.testutils.CommonTestUtils;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.streaming.api.graph.StreamGraph;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.test.util.AbstractTestBase;
+
+import org.junit.Assert;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+/**
+ * Tests an immediate checkpoint should be triggered right after all tasks reached the end of data.
+ */
+public class ImmediateCheckpointingAfterAllTasksFinishedITCase extends AbstractTestBase {
+    private static final int SMALL_SOURCE_NUM_RECORDS = 20;
+    private static final int BIG_SOURCE_NUM_RECORDS = 100;
+
+    private StreamExecutionEnvironment env;
+
+    @TempDir private java.nio.file.Path tmpDir;
+
+    @BeforeEach
+    public void setUp() {
+        env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        env.enableCheckpointing(Long.MAX_VALUE - 1);
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 4})
```

Comment:
Can you explain why we need to run this test with two different parallelism choices?

Note that most checkpoint tests in this package do not re-run tests with different parallelism. It is probably better to only introduce the extra complexity when we need it.

Reply:
The JobMaster receives RPCs from all subtasks of all vertices, only if all subtasks reached end-of-data, it can trigger the checkpoint. So I think we need to test the parallelism==1 as well as parallelism>1.

Reply:
hmm.. suppose we have verified that the test works with parallelism=4, can you help me understand why we need to additionally test parallelism=1?